### PR TITLE
Fix Squared Operator 

### PR DESCRIPTION
**Pull request by PhilipVinc**
Cherry picked from  #1065 so that i separate the two changes in two different PRs.

Computing the gradient of operators that use `nkjax.expect` instead of the covariance formula (such as `SquaredOperator`) also had a wrong factor of 2 for C->C models.
This is due to the weird way that Jax handles complex differentiation.

This PR now fixes it, and a test is added to check the gradient wrt finite differences.

Also, this PR moves out into a common test file the finite difference functions so that they can be used for other tests.

### CODE REVIEW

**Comment by gcarleo** [comment]
cool! was this affecting steady state solvers?

**Review by gcarleo — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,6 +24,7 @@
 * Fixed bug that prevented calling `.quantum_geometric_tensor` on `netket.vqs.ExactState`. [#1108](https://github.com/netket/netket/pull/1108)
 * Fixed bug where the gradient of `C->C` models (complex parameters, complex output) was computed incorrectly with `nk.vqs.ExactState`. [#1110](https://github.com/netket/netket/pull/1110)
 * Fixed bug where `QGTJacobianDense.state` and `QGTJacobianPyTree.state` would not correctly transform the starting point `x0` if `holomorphic=False`. [#1115](https://github.com/netket/netket/pull/1115)
+* The gradient of the expectation value obtained with `VarState.expect_and_grad` for `SquaredOperator`s was off by a factor of 2 in some cases, and wrong in others. This has now been fixed for models with real parameters. [#1065](https://github.com/netket/netket/pull/1065).
```
Comment (femtobit) [inline CHANGELOG.md:27]: The last sentence is ambiguous, because "This has now been fixed for models with real parameters" can imply both that complex parameters were never affected or that they were and this has not yet been fixed.

Comment (PhilipVinc) [reply CHANGELOG.md]: I just forgot to update the changelog.
Originally I had only fixed that case, but later on I fixed all cases...

**Comment by PhilipVinc** [comment]
Not really, it's just a factor of 2, but I'm trying to get those factors right.

**Comment by PhilipVinc** [comment]
The steady state does not go through this code path (yet. I'll open another PR to do that, as this code path is simpler, cleaner and faster)
